src/corecel/sys/MultiExceptionHandler.cc

```diff
@@ -0,0 +1,91 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2022 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file corecel/sys/MultiExceptionHandler.cc
+//---------------------------------------------------------------------------//
+#include "MultiExceptionHandler.hh"
+
+#include "corecel/Assert.hh"
+#include "corecel/io/Logger.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Throw the first exception and log all the rest.
+ */
+[[noreturn]] void log_and_rethrow_impl(MultiExceptionHandler&& exceptions)
+{
+    CELER_EXPECT(!exceptions.empty());
+    auto exc_vec = exceptions.release();
+
+    for (auto eptr_iter = exc_vec.begin() + 1; eptr_iter != exc_vec.end();
+         ++eptr_iter)
+    {
+        try
+        {
+            std::rethrow_exception(*eptr_iter);
+        }
+        catch (const std::exception& e)
+        {
+            CELER_LOG_LOCAL(critical) << "ignoring exception: " << e.what();
+        }
+        catch (...)
+        {
+            CELER_LOG_LOCAL(critical) << "ignoring exception of unknown type";
+        }
+    }
+
+    std::rethrow_exception(exc_vec.front());
+}
+} // namespace detail
+
+//---------------------------------------------------------------------------//
+/*!
+ * Terminate if destroyed without handling exceptions.
+ */
+MultiExceptionHandler::~MultiExceptionHandler()
+{
+    if (CELER_UNLIKELY(!exceptions_.empty()))
+    {
+        for (auto eptr : exceptions_)
+        {
+            try
+            {
+                std::rethrow_exception(eptr);
+            }
+            catch (const std::exception& e)
+            {
+                CELER_LOG_LOCAL(critical) << e.what();
+            }
+            catch (...)
+            {
+                CELER_LOG_LOCAL(critical) << "(unknown exception)";
+            }
+        }
+        CELER_LOG(critical) << "failed to clear exceptions from "
+                               "MultiExceptionHandler";
+        std::terminate();
+    }
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Thread-safe capture of the given exception.
+ */
+void MultiExceptionHandler::operator()(std::exception_ptr p)
+{
+#ifdef _OPENMP
+#    pragma omp critical
```

Comment:
One thing I learned the hard way — if you have multiple `pragma omp critical` directives in a codebase for different purposes, they effectively [share a mutex](https://www.openmp.org/spec-html/5.1/openmpsu99.html#x132-1420002.19.1:~:text=All%20critical%20constructs%20without%20a%20name%20are%20considered%20to%20have%20the%20same%20unspecified%20name.) if they don't have a name associated with them. While it's not an issue now because this is the only critical we have, to be safe it's always best to give a name:
```suggestion
#    pragma omp critical (MultiExceptionHandler)
```

Reply:
Wow! That's really subtle: the documentation is pretty ambiguous:
> The critical construct restricts execution of the associated structured block to a single thread at a time
but later in the details...
> All critical constructs without a name are considered to have the same unspecified name.

Reply:
Yeh - I found this out too when adding OpenMP to Shift...